### PR TITLE
Introduce shared auto-tune code.

### PR DESCRIPTION
**Pull request by magnumripper**
Only implemented in oldoffice format here. Just a PoC. See #2858

### CODE REVIEW

**Comment by magnumripper** [comment]
While adding 7z format I saw some caveats. Working on it. Meanwhile please do not add anything to this branch (because interface needs a change).

**Comment by jfoug** [comment]
Question.  Why only do this for OMP?

File: src/omp_autotune.c [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+			best_scale = scale;
+		}
+		else {
+			if (john_main_process && options.verbosity == VERB_MAX)
+				fprintf(stderr, " -\n");
+			no_progress++;
+		}
+
+		// Double each time
+		scale *= 2;
```
Comment (jfoug) [inline src/omp_autotune.c]: how about:
```c
scale += (scale)/2+1;
```
That would chose for values under 200k : 
1 2 4 7 11 17 26 40 61 92 139 209 314 472 709 1064 1597 2396 3595 5393 8090 12136 18205 27308 40963 61445 92168 138253
vs:
1 2 4 8 16 32 64 128 256 512 1024 2048 4096 8192 16384 32768 65536 131072

Just gives a bit more granularity.

Or possibly this expression:

```c
scale += scale < 5 ? 1 ? (scale/3)+1;
```
which gives this result (even better granularity):
1 2 3 4 5 8 12 18 26 36 50 68 92 124 167 224 300 402 538 719 960 1282 1711 2283 3046 4063 5419 7227 9638 12852 17138 22852 30471 40630 54175 72235 96315 128422 171231

File: src/omp_autotune.c [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+		do {
+			int count = this_kpc;
+
+			self->methods.crypt_all(&count, NULL);
+			crypts += count;
+		} while (sTimer_GetSecs(&timer) < 0.010);
+		sTimer_Stop(&timer);
+
+		// Release buffers
+		self->methods.done();
```
Comment (jfoug) [inline src/omp_autotune.c]: I was also calling clear_keys()   is there no reason to do this?

File: src/omp_autotune.c (hunk 1 above)
Comment (jfoug) [reply src/omp_autotune.c]: With a very limited testing (of the formats you already have done), the scale *=2 seems to work as good or better than either of the other 2 'finer' granular methods I proposed, even when I messed around with the ```if (cps >= (best_cps * 1.05))``` code a bit.

KISS method looks damn good ;)

**Comment by jfoug** [comment]
```
Testing: netlm, LM C/R [DES 32/64]... (2xOMP) =================================================================
==23708==ERROR: AddressSanitizer: global-buffer-overflow on address 0x55c653389707 at pc 0x55c652fa6f01 bp 0x7fff33299850 sp 0x7fff33299840
READ of size 1 at 0x55c653389707 thread T0
    #0 0x55c652fa6f00 in get_salt /cwd/src/NETLM_fmt_plug.c:282
    #1 0x55c652e0ab72 in omp_autotune /cwd/src/omp_autotune.c:39
    #2 0x55c652fa7526 in init /cwd/src/NETLM_fmt_plug.c:102
    #3 0x55c6531c9d43 in fmt_init /cwd/src/formats.c:80
    #4 0x55c6531eaaa5 in ldr_init_test_db /cwd/src/loader.c:1396
    #5 0x55c6531ae8e0 in benchmark_all /cwd/src/bench.c:807
    #6 0x55c6531dd23b in john_run /cwd/src/john.c:1617
    #7 0x55c6531deaea in main /cwd/src/john.c:2000
    #8 0x7f96b5d691c0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x211c0)
    #9 0x55c652d73bf9 in _start (/cwd/run/john+0x1c7bf9)
0x55c653389707 is located 57 bytes to the left of global variable '*.LC12' defined in 'NETLM_fmt_plug.c' (0x55c653389740) of size 10
  '*.LC12' is ascii string 'G3RG3P00!'
0x55c653389707 is located 6 bytes to the right of global variable '*.LC11' defined in 'NETLM_fmt_plug.c' (0x55c653389700) of size 1
  '*.LC11' is ascii string ''
SUMMARY: AddressSanitizer: global-buffer-overflow /cwd/src/NETLM_fmt_plug.c:282 in get_salt
Shadow bytes around the buggy address:
  0x0ab94a669290: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0ab94a6692a0: 00 00 00 00 00 f9 f9 f9 f9 f9 f9 f9 00 00 00 00
  0x0ab94a6692b0: 01 f9 f9 f9 f9 f9 f9 f9 00 00 01 f9 f9 f9 f9 f9
  0x0ab94a6692c0: 00 f9 f9 f9 f9 f9 f9 f9 06 f9 f9 f9 f9 f9 f9 f9
  0x0ab94a6692d0: 07 f9 f9 f9 f9 f9 f9 f9 00 02 f9 f9 f9 f9 f9 f9
=>0x0ab94a6692e0:[01]f9 f9 f9 f9 f9 f9 f9 00 02 f9 f9 f9 f9 f9 f9
  0x0ab94a6692f0: 05 f9 f9 f9 f9 f9 f9 f9 00 00 00 00 00 00 01 f9
  0x0ab94a669300: f9 f9 f9 f9 00 02 f9 f9 f9 f9 f9 f9 00 00 01 f9
  0x0ab94a669310: f9 f9 f9 f9 00 00 00 00 00 00 00 00 00 01 f9 f9
  0x0ab94a669320: f9 f9 f9 f9 00 02 f9 f9 f9 f9 f9 f9 00 00 00 00
  0x0ab94a669330: 00 00 00 00 00 01 f9 f9 f9 f9 f9 f9 00 07 f9 f9
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==23708==ABORTING
```

**Comment by jfoug** [comment]
NetLM needs prepare called, I think.  Or else, we need to validate that all formats have test hash 0 in properly prepared format.

File: src/omp_autotune.c (hunk 2 above)
Comment (magnumripper) [reply src/omp_autotune.c]: Right, that needs to be added

**Comment by magnumripper** [comment]
> Question. Why only do this for OMP?

Good question! We can look into that too.

**Comment by magnumripper** [comment]
NETLM and others are fixed in my latest code but it's not quite finished yet. It also adds tuning for most expensive salt of all test vectors or - in case we have real hashes loaded - the most expensive salt of those (similar to OpenCL autotune). I will probably not finish it or push anything new here in 48h or so (guess why 🎆 )

**Comment by magnumripper** [comment]
In order to tune for actual loaded hashes and/or heaviest cost, I had to change interface for `omp_autotune()`. We need to get a pointer to `self` (format struct) as well as the `db`. This is not sweet but works:
```diff
 static void init(struct fmt_main *self)
 {
 #ifdef _OPENMP
-       int threads = omp_get_max_threads();
-
-       if (threads > 1) {
-               self->params.min_keys_per_crypt *= threads;
-               threads *= OMP_SCALE;
-               self->params.max_keys_per_crypt *= threads;
-       }
+       omp_autotune(self, NULL);
 #endif
        saved_key = mem_calloc(self->params.max_keys_per_crypt,
                               sizeof(*saved_key));
@@ -122,6 +108,13 @@ static void init(struct fmt_main *self)
        output = mem_calloc(self->params.max_keys_per_crypt, sizeof(*output));
 }
 
+static void reset(struct db_main *db)
+{
+#if defined (_OPENMP)
+       omp_autotune(NULL, db);
+#endif
+}
+
 static void done(void)
 {
        MEM_FREE(output);
```
The call from `init()` merely saves off the `self` pointer and returns number of actual threads. The call from `reset()` will actually tune a scale. If real hashes are loaded, they will be used for tuning instead of test vectors. Otherwise the test vectors are used.

I'm actually not quite sure whether we should tune for heaviest or lightest cost? An experienced user will usually not run with very different costs at once though. Lightest cost would sometimes mean a much faster autotune.

@solardiz any thoughts (this is still just an experiment BTW)?

**Comment by magnumripper** [comment]
I think this is about as good as this gets. Mostly KISS, mostly working as a champ. Some quirks. This really reveals formats like 7z where the speed difference between fast and slow salts are gigantic. We have no option other than to benchmark for slowest salt, or self-test does take ages. OTOH we have the exat same problem with any scheme of manual tuning too.

Not sure whether we should merge this or just consider it an interesting experiment.

**Comment by magnumripper** [comment]
Brand new version, less intrusive. Keeps the OMP_SCALE values and uses them by default.

More to come.

**Comment by magnumripper** [comment]
Rebased a cleaned-up version. Now formats.h in turn sources omp_autotune.h so most formats merely have this change:
```diff
diff --git a/src/AzureAD_fmt_plug.c b/src/AzureAD_fmt_plug.c
index 35f5b8011..b6a64bf72 100644
--- a/src/AzureAD_fmt_plug.c
+++ b/src/AzureAD_fmt_plug.c
@@ -80,13 +80,7 @@ static uint32_t (*crypt_out)[BINARY_SIZE / sizeof(uint32_t)];
 static void init(struct fmt_main *self)
 {
 #ifdef _OPENMP
-       int threads = omp_get_max_threads();
-
-       if (threads > 1) {
-               self->params.min_keys_per_crypt *= threads;
-               threads *= OMP_SCALE;
-               self->params.max_keys_per_crypt *= threads;
-       }
+       omp_autotune(self, OMP_SCALE);
 #endif
        saved_key = mem_calloc_align(self->params.max_keys_per_crypt,
                                     sizeof(*saved_key), MEM_ALIGN_WORD);
```

**Comment by magnumripper** [comment]
I simply had to merge this before @kholia forced me to rebase yet another again 😄 

All tests passed and behavior isn't changed unless you ask for it.

**Pull request merged**